kmsv2: re-use DEK while key ID is unchanged

enj opened the pull request:
This change updates KMS v2 to not create a new DEK for every
encryption.  Instead, we re-use the DEK while the key ID is stable.

Specifically:

We no longer use a random 12 byte nonce per encryption.  Instead, we
use both a random 4 byte nonce and an 8 byte nonce set via an atomic
counter.  Since each DEK is randomly generated and never re-used,
the combination of DEK and counter are always unique.  Thus there
can never be a nonce collision.  AES GCM strongly encourages the use
of a 12 byte nonce, hence the additional 4 byte random nonce.  We
could leave those 4 bytes set to all zeros, but there is no harm in
setting them to random data (it may help in some edge cases such as
live VM migration).

If the plugin is not healthy, the last DEK will be used for
encryption for up to three minutes (there is no difference on the
behavior of reads which have always used the DEK cache).  This will
reduce the impact of a short plugin outage while making it easy to
perform storage migration after a key ID change (i.e. simply wait
ten minutes after the key ID change before starting the migration).

The DEK rotation cycle is performed in sync with the KMS v2 status
poll thus we always have the correct information to determine if a
read is stale in regards to storage migration.

Signed-off-by: Monis Khan <mok@microsoft.com>

/kind feature
/kind api-change

Fixes: #90856

xref: #87438

```release-note
The API server now re-uses data encryption keys while the kms v2 plugin's key ID is stable.  Data encryption keys are still randomly generated on server start but an atomic counter is used to prevent nonce collisions.
```

enj committed:
no-op: drop baseTransformerFunc indirection

Signed-off-by: Monis Khan <mok@microsoft.com>

enj committed:
kmsv2: re-use DEK while key ID is unchanged

This change updates KMS v2 to not create a new DEK for every
encryption.  Instead, we re-use the DEK while the key ID is stable.

Specifically:

We no longer use a random 12 byte nonce per encryption.  Instead, we
use both a random 4 byte nonce and an 8 byte nonce set via an atomic
counter.  Since each DEK is randomly generated and never re-used,
the combination of DEK and counter are always unique.  Thus there
can never be a nonce collision.  AES GCM strongly encourages the use
of a 12 byte nonce, hence the additional 4 byte random nonce.  We
could leave those 4 bytes set to all zeros, but there is no harm in
setting them to random data (it may help in some edge cases such as
live VM migration).

If the plugin is not healthy, the last DEK will be used for
encryption for up to three minutes (there is no difference on the
behavior of reads which have always used the DEK cache).  This will
reduce the impact of a short plugin outage while making it easy to
perform storage migration after a key ID change (i.e. simply wait
ten minutes after the key ID change before starting the migration).

The DEK rotation cycle is performed in sync with the KMS v2 status
poll thus we always have the correct information to determine if a
read is stale in regards to storage migration.

Signed-off-by: Monis Khan <mok@microsoft.com>

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -296,13 +296,94 @@ func (h *kmsv2PluginProbe) check(ctx context.Context) error {
+// between status and encrypt only results in log statements - the system attempts to coasts otherwise.
+// In practice this means that reads will coast indefinitely on the last valid state whereas writes will coast
+// until the DEK is considered to be expired.
+func (h *kmsv2PluginProbe) attemptToRotateDEK(ctx context.Context, statusKeyID string) error {
```
TODO: add unit tests for this.

enj commented:
/triage accepted

enj commented:
/hold

Needs API review before merging.

enj commented:
/priority important-soon

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -55,16 +56,18 @@ import (
+	secretboxTransformerPrefixV1     = "k8s:enc:secretbox:v1:"
+	kmsTransformerPrefixV1           = "k8s:enc:kms:v1:"
+	kmsTransformerPrefixV2           = "k8s:enc:kms:v2:"
+	kmsPluginHealthzPositiveInterval = 1 * time.Minute
+	kmsPluginHealthzNegativeInterval = 10 * time.Second
+	kmsPluginDEKReuseInterval        = 3 * time.Minute
```
Important.

mikedanese replied:
What's stopping us from making this 24 hours or more (or removing it altogether)?

Also something like "writeDEKMaxAge" might be a clearer name.

enj replied:
The gist is that it makes it impossible to have a "passive" storage migration.  With the current approach, you update the key ID your plugin is using and then can simply schedule a migration for `10+N` minutes later where `N` is how long it takes your plugin to pick up new config.  At that point, you are guaranteed to either migrate to the new key or get errors during the migration.  If the API server coasted forever on the last DEK, you would need to actively check if it had observed the new key ID before starting a migration - otherwise it could keep using the old DEK and your storage migration would not do what you thought it did.  I think we can iterate on making this approach better but for now I think we need to stick with the safe approach.

enj commented on staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go:
```diff
@@ -23,14 +23,23 @@ import (
 	"crypto/aes"
```
Important (entire file).

enj commented on staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go:
```diff
@@ -54,44 +56,65 @@ const (
 	encryptedDEKMaxSize = 1 * 1024 // 1 kB
 	// cacheTTL is the default time-to-live for the cache entry.
-	cacheTTL = 1 * time.Hour
+	// this allows the cache to grow to an infinite size for up to a day.
+	// this is meant as a temporary solution until the cache is re-written to not have a TTL.
+	// there is unlikely to be any meaningful memory impact on the server
+	// because the cache will likely never have more than a few thousand entries
+	// and each entry is roughly ~200 bytes in size.
+	cacheTTL = 24 * time.Hour
```
Important.

mikedanese replied:
It would be good to export a metric for cache size.

enj replied:
Will do this in a follow-up (have a re-work of the cache in progress).

enj reviewed (commented):
@liggitt marked the really important bits for extra careful review.

enj commented:
/assign @aramase @liggitt

aramase commented on staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go:
```diff
@@ -43,44 +52,75 @@ import (
+// use because it makes assumptions about the key underlying the input block cipher.  Specifically,
+// it uses a 96-bit nonce where the first 32 bits are random data and the remaining 64 bits are
+// a monotonically incrementing atomic counter.  This means that the key must be randomly generated
+// on process startup and must be never be used for encryption outside the lifetime of the process.
```
```suggestion
// on process startup and must never be used for encryption outside the lifetime of the process.
```

aramase commented on staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go:
```diff
@@ -43,44 +52,75 @@ import (
+
+		incrementingNonce := nonce.Add(1)
+		if incrementingNonce == 0 {
+			klog.Fatal("aes-gcm detected nonce overflow") // this should never happen, and is unrecoverable if it does
```
`Fatal` can't be used with structured logging, so we can panic here?

enj replied:
I'll update this to `ErrorS` + `os.Exit` so it always crashes the program (`panic` would get caught and subsequent requests would have nonce re-use).

mikedanese replied:
Fatal is nice because it includes a stack trace. Given this condition is very unexpected, a stack trace might come in handy.

aramase reviewed (commented):
Still reviewing the tests and changes in `kmsv2` package

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -296,13 +298,86 @@ func (h *kmsv2PluginProbe) check(ctx context.Context) error {
+		return nil
+	}
+
+	return fmt.Errorf("failed to rotate DEK uid=%q, errState=%v, errGen=%v, statusKeyID=%q, encryptKeyID=%q, stateKeyID=%q, expirationTimestamp=%s",
```
> this could cause the DEK to expire and be unavailable in now + `kmsPluginDEKReuseInterval` depending on when it was set last time

Which is exactly what we want?

aramase replied:
Updated the comment. The way it is now, it can actually expire before `now + kmsPluginDEKReuseInterval` because we haven't set the expiration time after the failure?

mikedanese commented on staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go:
```diff
@@ -43,44 +52,75 @@ import (
+
+	// even at one million encryptions per second, this counter is enough for half a million years
+	// using this struct avoids alignment bugs: https://pkg.go.dev/sync/atomic#pkg-note-BUG
+	var nonce atomic.Uint64
```
Is there a downside to making this counter global to the process? I'm worried about calling NewGCMTransformerWithUniqueKeyUnsafe() on the same block twice.

enj replied:
I think I can make that work.

mikedanese replied:
I retract my comment if you move block.Cipher generation into this function though.

mikedanese commented on staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go:
```diff
@@ -43,44 +52,75 @@ import (
+	// even at one million encryptions per second, this counter is enough for half a million years
+	// using this struct avoids alignment bugs: https://pkg.go.dev/sync/atomic#pkg-note-BUG
+	var nonce atomic.Uint64
+	nonceFunc := func(b []byte) error {
```
Optional,  consider extracting this to a struct:

```go
type nonceGenerator atomic.Uint64

func (...) next() []byte
```

mikedanese commented on staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go:
```diff
@@ -43,44 +52,75 @@ import (
+	nonceFunc := func(b []byte) error {
+		// we only need 8 bytes to store our 64 bit incrementing nonce
+		// instead of leaving the unused bytes as zeros, set those to random bits
+		// this mostly protects us from weird edge cases like a VM restore that rewinds our atomic counter
```
It also provides some protection if we accidentally calling NewGCMTransformerWithUniqueKeyUnsafe() on the same block multiple times (thereby resetting the counter).

mikedanese commented on staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go:
```diff
@@ -43,44 +52,75 @@ import (
+// on process startup and must be never be used for encryption outside the lifetime of the process.
+// Unlike NewGCMTransformer, this function is immune to the birthday attack and thus the key can
+// be used indefinitely without rotation.
+func NewGCMTransformerWithUniqueKeyUnsafe(block cipher.Block) (value.Transformer, error) {
```
Brainstorming a bit: how easy would it be to tag the cipher.Block implementation with a private function e.g. `_, ok := block.(interface{ derivedInProcess() })` then test for that here? Would that make this method substantially less error prone?

enj replied:
I am going to refactor this to not take a `block` as input and instead generate a random block on every invocation.  That is the only way to use this function safely anyway.  For the decryption path I will use the other function to create the transformer.

ritazh commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -296,13 +298,86 @@ func (h *kmsv2PluginProbe) check(ctx context.Context) error {
+	metrics.RecordKeyIDFromStatus(h.name, statusKeyID)
+
+	// we do not check ValidateEncryptCapability here because it is fine to re-use an old key
+	// that was mark as expired during an unhealthy period.  as long as the key ID matches
```
```suggestion
	// that was marked as expired during an unhealthy period.  as long as the key ID matches
```

ritazh commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -296,13 +298,86 @@ func (h *kmsv2PluginProbe) check(ctx context.Context) error {
+	}
+
+	if len(state.EncryptedDEK) == 0 {
+		return envelopekmsv2.State{}, fmt.Errorf("got unexpected empty EDEK")
```
```suggestion
		return envelopekmsv2.State{}, fmt.Errorf("got unexpected empty EncryptedDEK")
```

enj commented:
@ritazh @mikedanese @aramase I have addressed all comments, here is the [diff](https://github.com/kubernetes/kubernetes/compare/e9b68a26e78fe6194141302727cd09a83dc4b3a5..0187a8cd3ff2ba80f857114b748acc60aa9c3b5d?w=1).  I will rebase on master to fix the merge conflict.

valaparthvi commented:
Hi @enj, I am from the Bug Triage Team for K8s release v1.27. I see this PR is in actively worked up, so I assume it is on track for this release.

However, please note that Code Freeze begins on March 15, after which only critical and release blocker PR/issues will be accepted. Please make sure this PR goes in by March 15.

dims commented:
/skip

liggitt commented on staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go:
```diff
@@ -43,44 +53,135 @@ import (
+// a monotonically incrementing atomic counter.  This means that the key must be randomly generated
+// on process startup and must never be used for encryption outside the lifetime of the process.
+// Unlike NewGCMTransformer, this function is immune to the birthday attack and thus the key can
+// be used "indefinitely" without rotation.  Specifically, cryptographic wear out of AES-GCM with
```
doc:
```suggestion
// be used for 2^64-1 writes without rotation.  Specifically, cryptographic wear out of AES-GCM with
```

liggitt commented on staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go:
```diff
@@ -43,44 +53,135 @@ import (
+}
+
+func randomNonce(b []byte) error {
+	_, err := rand.Read(b)
```
does this return an error if fewer than len(b) bytes are read?

enj replied:
[Yes](https://github.com/golang/go/blob/2cbab4e98c6091f5fb6cb73bdebfe328793da388/src/crypto/rand/rand.go#L23).

liggitt commented on staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go:
```diff
@@ -43,44 +53,135 @@ import (
 
-// NewGCMTransformer takes the given block cipher and performs encryption and decryption on the given
-// data.
-func NewGCMTransformer(block cipher.Block) value.Transformer {
```
do we have a test that captures data encrypted by a previously released version of this and ensures the current version can still decrypt it?

enj replied:
I will add an explicit one.  Note that I didn't not touch decryption so it should be fine since existing tests are passing.

liggitt replied:
> Note that I didn't not touch decryption

I know that was the intent... enough stuff moved around that I'm having trouble eyeballing it and being 100% certain

enj replied:
> > do we have a test that captures data encrypted by a previously released version of this and ensures the current version can still decrypt it?
> 
> I didn't see a fixture of obviously encrypted data... did I miss this test?

[TestGCMLegacyDataCompatibility](https://github.com/kubernetes/kubernetes/pull/116155/files#diff-e8f886d4a6f00f2bcc161c87a21332c574f1092958f8be274caf6e2287cad95dR189-R190)

https://github.com/kubernetes/kubernetes/blob/02031085033a88f49a190ce45c79459edb5adca4/staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes_test.go#L189-L190

> > Note that I didn't not touch decryption
> 
> I feel like that is a good representation of my brain after reviewing this PR

😂 😭

liggitt replied:
> // recorded output from NewGCMTransformer at commit 3b1fc60d8010dd8b53e97ba80e4710dbb430beee

thanks, verified that package hasn't changed since v1.26.0, so it's representative of released state

liggitt commented on staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go:
```diff
@@ -43,44 +53,135 @@ import (
+// around plaintext length because all stored items are small (less than 2 MB).  To prevent the
+// chance of the block cipher being accidentally re-used, it is not taken in as input.  Instead,
+// a new random key is generated and returned on every invocation of this function.  This key is
+// used as the input to the block cipher.
```
> Instead, a new random key is generated and returned on every invocation of this function.
> This key is used as the input to the block cipher.

doc: ... and can be passed to `NewGCMTransformer` to construct a transformer capable of decrypting values encrypted by this transformer?

liggitt commented on staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go:
```diff
@@ -43,44 +53,135 @@ import (
+// chance of the block cipher being accidentally re-used, it is not taken in as input.  Instead,
+// a new random key is generated and returned on every invocation of this function.  This key is
+// used as the input to the block cipher.
+func NewGCMTransformerWithUniqueKeyUnsafe() (value.Transformer, []byte, error) {
```
nit: but "Unsafe" freaks some people out... wondering if there's a better suffix like "Restricted"

enj replied:
But I do want to freak people out 🙂

This code is one trivial mistake away from catastrophic failure...

liggitt commented on staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go:
```diff
@@ -43,44 +53,135 @@ import (
-	block cipher.Block
-}
+func NewGCMTransformer(block cipher.Block) (value.Transformer, error) {
+	aead, err := cipher.NewGCM(block)
```
Is `aead` safe to reuse and threadsafe? previously we constructed a local one within every call to TransformFromStorage / TransformToStorage, now we reuse the same one for the whole instance. It's not clear to me that is safe.

enj replied:
Yup, a member of the Go team confirmed that it was go routine safe / stateless (the std lib code is hard to read because it is mostly assembly).

enj replied:
In `master` we already rely on the assumption that `cipher.Block` is go routine safe, so this does not really change anything:

https://github.com/kubernetes/kubernetes/blob/fcf5d23e6818ccfbc92f332af82f31661f8d4f88/staging/src/k8s.io/kms/pkg/encrypt/aes/aes.go#L50-L60

From the discussion in https://github.com/golang/go/issues/41689 (which was a follow up to https://github.com/golang/go/issues/25882), `cipher.NewGCM` is also go routine safe (and it is stateless).

Even without the race detector, `TestGCMUnsafeNonceGen` would detect data corruption (so at least we know it is okay on `linux/amd64` and `darwin/arm64`).

liggitt replied:
I did trace that we were relying on the aes.NewCipher implementation of cipher.Block to be threadsafe, so I agree this PR doesn't change that.

From what I can tell, it does introduce reliance on cipher.NewGCM being threadsafe.

I agree the comments on those go issues and from the go maintainer indicate it is threadsafe, but I would feel *much* better if that was tested/documented on cipher.NewGCM and aes.NewCipher. Perhaps a post-code freeze contribution to go1.21 to solidify what we're relying on here?

liggitt commented on staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go:
```diff
@@ -43,44 +53,135 @@ import (
-	nonceSize := aead.NonceSize()
-	if len(data) < nonceSize {
-		return nil, false, fmt.Errorf("the stored data was shorter than the required size")
+	block, err := aes.NewCipher(key)
```
same question about whether `block` is safe to reuse and threadsafe

enj replied:
I can check but we already use this in other places that require it to be go routine safe.

enj replied:
Confirmed that this is [safe](https://github.com/golang/go/blob/b6f29d23654eba09209db00704ccef1f01a60c71/src/crypto/aes/block.go#L43-L133) as well (`xk` is only read from).

liggitt commented on staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go:
```diff
@@ -43,44 +53,135 @@ import (
-func (t *gcm) TransformToStorage(ctx context.Context, data []byte, dataCtx value.Context) ([]byte, error) {
-	aead, err := cipher.NewGCM(t.block)
+func newGCMTransformerWithUniqueKeyUnsafe(block cipher.Block, nonceGen *nonceGenerator) (value.Transformer, error) {
+	aead, err := cipher.NewGCM(block)
```
not sure if @mikedanese has reviewed the latest state of this closely, but I'd really like a second or third set of eyes on this (especially the `cipher.NewGCM(aes.NewCipher(key))` wiring in the new code path for correctness)

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -310,30 +331,106 @@ func (h *kmsv2PluginProbe) check(ctx context.Context) error {
+		errs = append(errs, fmt.Errorf("got invalid KMSv2 KeyID %q: %w", response.KeyID, err))
+	} else {
+		metrics.RecordKeyIDFromStatus(h.name, response.KeyID)
+		errs = append(errs, h.rotateDEKOnKeyIDChange(ctx, response.KeyID, string(uuid.NewUUID())))
```
I will add a comment to make this more clear.

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -60,11 +61,38 @@ const (
+	// amount of plugin downtime.
+	//
+	// With the current approach, a user can update the key ID their plugin
+	// is using and then can simply schedule a migration for 10+N minutes
```
I was thinking that if the plugin broke / somehow the DEK loop stopped running, you have a 3 minute window in which you won't notice, so you probably want to wait a minimum of 5 minutes.  So just double the minimum and you get 10.  I can write that out or pick a different number?

liggitt replied:
hmm... I thought the logic meant we set the expiration to 3 minutes from just before the last time we successfully observed the key id matched... so isn't 3 minutes from when your KMS plugin starts reporting the new keyid when you can start storage migration? not sure padding that and then doubling it is necessary... I would just document the time offsets this system has internally, and leave any padding up to the caller.

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -60,11 +61,38 @@ const (
+	// storage migration would not do what they thought it did.
+	kmsv2PluginHealthzPositiveInterval = 1 * time.Minute
+	kmsv2PluginHealthzNegativeInterval = 10 * time.Second
+	kmsv2PluginWriteDEKMaxAge          = 3 * time.Minute
```
I am open to whatever variable name is most clear.  A single DEK is used until API server restart/key ID change as discussed in the last SIG Auth meeting.

https://github.com/kubernetes/kubernetes/pull/116155#discussion_r1130232693

enj commented on staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go:
```diff
@@ -103,8 +125,22 @@ func (t *envelopeTransformer) TransformFromStorage(ctx context.Context, data []b
 
-	// Look up the decrypted DEK from cache or Envelope.
-	transformer := t.cache.get(encryptedObject.EncryptedDEK)
+	state, err := t.stateFunc() // no need to call state.ValidateEncryptCapability on reads
+	if err != nil {
+		return nil, false, err
+	}
+
+	// start with the assumption that the current write transformer is also the read transformer
+	transformer := state.Transformer
+
+	// if the current write transformer is not what was used to encrypt this data, check in the cache
+	// TODO: consider marking this as a stale read to support DEK defragmentation
+	if subtle.ConstantTimeCompare(state.EncryptedDEK, encryptedObject.EncryptedDEK) != 1 {
```
Oh and the current cache expires entries on a TTL, and it felt silly for the server's own current write DEK to ever expire out.

liggitt replied:
eh... ok... feels like a comparison and check that will turn out to be wrong for most items after a while in most clusters

at least drop the constant time comparison?

enj commented on staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/cache.go:
```diff
@@ -66,7 +68,7 @@ func (c *simpleCache) set(key []byte, transformer value.Transformer) {
 		panic("transformer must not be nil")
 	}
-	c.cache.Set(c.keyFunc(key), transformer, c.ttl)
+	c.cache.Set(c.keyFunc(key), &decryptOnlyTransformer{transformer: transformer}, c.ttl)
```
My thought was that the transformers that are unsafe for encryption are only used through the cache, hence I wrapped them.  I can, in addition to wrapping them, strip down the interface to only allow decryption.

liggitt replied:
if the cache should only contain decrypting transformers, I guess narrow the interface of the type stored/returned from the cache

wrapping inside the cache is a little strange... if we still feel like we need to wrap, I guess do that when constructing the decrypt-only transformers?

enj commented on staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go:
```diff
@@ -123,90 +159,74 @@ func (t *envelopeTransformer) TransformFromStorage(ctx context.Context, data []b
-	transformer := t.baseTransformerFunc(block)
+	// this is compatible with NewGCMTransformerWithUniqueKeyUnsafe for decryption
+	// it would use random nonces for encryption but we never do that
+	transformer, err := aestransformer.NewGCMTransformer(block)
```
Is this sufficient:

https://github.com/kubernetes/kubernetes/pull/116155#discussion_r1132907618

Or should I wrap before putting in the cache?

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -291,15 +319,8 @@ func (h *kmsv2PluginProbe) check(ctx context.Context) error {
-	}
 
-	if err := isKMSv2ProviderHealthy(h.name, p); err != nil {
+	if err := h.isKMSv2ProviderHealthyAndMaybeRotateDEK(ctx, p); err != nil {
```
I will add a comment.  Note that they were already load bearing for KMS v2 because the server cannot perform reads without knowing the key ID.

enj commented:
@liggitt [comments addressed](https://github.com/kubernetes/kubernetes/compare/6bcfa293c02860f15978c0f150ded5546d1289fc..02031085033a88f49a190ce45c79459edb5adca4?w=1).

/hold cancel

enj commented on staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go:
```diff
@@ -43,44 +53,137 @@ import (
+	aead, err := cipher.NewGCM(block)
 	if err != nil {
 		return nil, err
 	}
```
Added, and `TestGCMDataStable` cover this as well:

https://github.com/kubernetes/kubernetes/blob/02031085033a88f49a190ce45c79459edb5adca4/staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes_test.go#L47-L66

enj commented:
@aramase @liggitt comments addressed, here is the [diff](https://github.com/kubernetes/kubernetes/compare/02031085033a88f49a190ce45c79459edb5adca4..832d6f0e19f13b9dd22b1fe9d705817e9e64f4f1?w=1).

liggitt commented:
/lgtm
/approve
